File: leetcode/11.盛最多水的容器.js
/*
 * @lc app=leetcode.cn id=11 lang=javascript
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    let left = 0;
    let right = height.length - 1;
    let maxArea = 0;
    while( right > left ){
        let area = 0;
        let dis = right - left;
        if( height[left] > height[right] ){
            area = dis * height[right];
            right--;
        }else{
            area = dis * height[left];
            left++;
        }
        if( area > maxArea ){
            maxArea = area;
        }
    }
    return maxArea;
};

//暴力法
var maxArea1 = function(height) {
    let maxArea = 0;
    for (let i = 0; i < height.length-1; i++) {
        for (let j = i; j < height.length; j++) {
            let area = Math.min( height[i], height[j] ) * (j - i);
            if( area > maxArea ){
                maxArea = area;
            }
        }
    }
    return maxArea;
};
// @lc code=end

File: leetcode/11.盛最多水的容器.go
/*
 * @lc app=leetcode.cn id=11 lang=golang
 *
 * [11] 盛最多水的容器
 */

// @lc code=start
func maxArea(height []int) int {
	ret, left, right := 0, 0, len(height)-1
	for left < right {
		if height[left] < height[right] {
			ret = max(ret, (right-left)*min(height[left], height[right]))
			left++
		} else {
			ret = max(ret, (right-left)*min(height[left], height[right]))
			right--
		}
	}
	return ret
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

// @lc code=end

